File: FlinkScalaCode/src/main/scala/com/wubaibao/flinkscala/code/chapter6/partitions/PartitionCustomTest.scala
package com.wubaibao.flinkscala.code.chapter6.partitions

import org.apache.flink.api.common.functions.Partitioner
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}

/**
 * Flink自定义分区器
 */
object PartitionCustomTest {
  def main(args: Array[String]): Unit = {

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //导入隐式转换
    import org.apache.flink.streaming.api.scala._
    val ds1: DataStream[String] = env.socketTextStream("node5", 9999)
    val ds2: DataStream[(String, Int)] = ds1.map(one => {
      val arr: Array[String] = one.split(",")
      (arr(0), arr(1).toInt)
    })

    val result: DataStream[(String, Int)] = ds2.partitionCustom(new Partitioner[String] {
      override def partition(key: String, numPartitions: Int): Int = {
        key.hashCode % numPartitions
      }
    }, _._1)

    result.print()
    env.execute()
  }
}

File: FlinkJavaCode/src/main/java/com/wubaibao/flinkjava/code/chapter6/partitions/PartitionCustomTest.java
package com.wubaibao.flinkjava.code.chapter6.partitions;

import org.apache.flink.api.common.functions.Partitioner;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.RestOptions;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * Flink自定义分区器实现
 */
public class PartitionCustomTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        /**
         * Socket中输入数据如下格式数据
         * a,1
         * b,2
         * a,3
         * b,4
         * c,5
         */
        DataStreamSource<String> ds1 = env.socketTextStream("node5", 9999);
        SingleOutputStreamOperator<Tuple2<String, Integer>> ds2 = ds1.map(one -> {
            String[] arr = one.split(",");
            return Tuple2.of(arr[0], Integer.valueOf(arr[1]));
        }).returns(Types.TUPLE(Types.STRING, Types.INT));

        DataStream<Tuple2<String, Integer>> result = ds2.partitionCustom(new Partitioner<String>() {
            @Override
            public int partition(String key, int numPartitions) {
                return key.hashCode() % numPartitions;
            }
        }, new KeySelector<Tuple2<String, Integer>, String>() {
            @Override
            public String getKey(Tuple2<String, Integer> tp) throws Exception {
                return tp.f0;
            }
        });

        result.print();
        env.execute();

    }
}
